src/shared/sol-iio.c

```diff
@@ -0,0 +1,990 @@
+            goto error;
+        }
+
+        if (buffer_size != 0)
+            set_buffer_size(device, buffer_size);
+
+        if (!device->manual_triggering) {
+            SOL_WRN("No 'trigger_now' file on device%d current trigger. "
+                "It won't be possible to manually trigger a reading on device",
+                device->device_id);
+        }
+
+        if (!setup_device_reader(device)) {
+            SOL_WRN("Could not setup device%d reading", device->device_id);
+            goto error;
+        }
+        device->buffer_enabled = true;
+    } else {
+        /* buffer_size == -1 means that user doesn't want to use the buffer */
+        device->buffer_enabled = false;
+        if (!set_buffer_enabled(device, false)) {
+            SOL_WRN("Could not enable buffer for device");
+            goto error;
+        }
+    }
+
+    SOL_DBG("iio device created. device%d - buffer_enabled: %d - manual_trigger: %d"
+        " - trigger_name: %s - trigger_id: %d", device->device_id,
+        device->buffer_enabled, device->manual_triggering, device->trigger_name,
+        device->trigger_id);
+
+    return device;
+
+error:
+    sol_iio_close(device);
+
+    return NULL;
+}
+
+static bool
+enable_channel_scan(struct sol_iio_channel *channel)
+{
+    struct sol_iio_device *device = channel->device;
+    char path[PATH_MAX + 1];
+    int current_value;
+
+    if (!craft_filename_path(path, sizeof(path), CHANNEL_SCAN_ENABLE_PATH,
+        device->device_id, channel->name)) {
+
+        return false;
+    }
+
+    /* First, check if not already enabled */
+    sol_util_read_file(path, "%d", &current_value);
+    if (current_value != 1)
+        return write_file(path, "%d", 1);
+
+    return true;
+}
+
+static bool
+read_channel_scan_info(struct sol_iio_channel *channel)
+{
+    struct sol_iio_device *device = channel->device;
+    char path[PATH_MAX + 1];
+    char *type, *original_type;
+    int len;
+    bool crafted;
+
+    crafted = craft_filename_path(path, sizeof(path), CHANNEL_SCAN_TYPE_PATH,
+        device->device_id, channel->name);
+    if (!crafted || sol_util_read_file(path, "%ms", &type) < 0)
+        return false;
+
+    original_type = type;
+    /* Form of type is [be|le]:[s|u]bits/storagebits[>>shift] */
+    if (strstr(type, "be:")) {
+        channel->little_endian = false;
+        type += 3;
+    } else if (strstr(type, "le:")) {
+        channel->little_endian = true;
+        type += 3;
+    } else
+        channel->little_endian = true;
+
+    if (type[0] == 's') {
+        channel->is_signed = true;
+        type++;
+    } else if (type[0] == 'u') {
+        channel->is_signed = false;
+        type++;
+    } else
+        channel->is_signed = true; /* TODO check if signed is the default */
+
+    channel->shift = 0;
+    len = sscanf(type, "%d/%d>>%d", &channel->bits, &channel->storagebits, &channel->shift);
+
+    free(original_type);
+
+    if (len < 0)
```

Comment:
Only return `true` if `len` is `3`, as not all items might have been correctly converted.